test_runner/src/main/kotlin/ftl/reports/outcome/BillableMinutes.kt

```diff
@@ -0,0 +1,46 @@
+package ftl.reports.outcome
+
+import com.google.api.services.toolresults.model.Step
+import com.google.api.services.toolresults.model.StepDimensionValueEntry
+import ftl.android.AndroidCatalog
+import ftl.util.billableMinutes
+import kotlin.math.min
+
+data class BillableMinutes(
+    val virtual: Long = 0,
+    val physical: Long = 0
+)
+
+fun List<Step>.calculateAndroidBillableMinutes(
+    projectId: String,
+    timeoutValue: Long
+): BillableMinutes =
+    groupByDeviceType(projectId).run {
+        BillableMinutes(
+            virtual = get(true)?.sumBillableMinutes(timeoutValue) ?: 0,
+            physical = get(false)?.sumBillableMinutes(timeoutValue) ?: 0
+        )
+    }
+
+private fun List<Step>.groupByDeviceType(projectId: String) =
+    groupBy {
+        AndroidCatalog.isVirtualDevice(
+            it.deviceModel(),
+            projectId
+        )
+    }
+
+private fun List<Step>.sumBillableMinutes(timeout: Long) =
+    mapNotNull { step ->
+        step.getBillableSeconds(default = timeout)
+    }.map {
+        billableMinutes(it)
+    }.sum()
+
+private fun Step.getBillableSeconds(default: Long) =
+    testExecutionStep?.testTiming?.testProcessDuration?.seconds?.let {
+        min(it, default)
+    }
+
+operator fun List<StepDimensionValueEntry>.get(key: String) =
```

Comment:
In `Util.kt` we have `dimensionValue["Model"]` --  `dimensionValue` is nullable. How about making this operator null-proof?

Reply:
So what should be default return value for this function?

Reply:
We do not have to add default value, just make it safe call, if `get` returns `null` we are prepared for it
```kotlin
internal fun Step.deviceModel() = dimensionValue["Model"]
    ?.value.orUnknown()
```

Reply:
done